universal-application-tool-0.0.1/app/services/applicant/Block.java

```diff
@@ -1,65 +1,67 @@
-public final class Block {
+@AutoValue
+public abstract class Block {
+
+  public static Block create(
+      long id, BlockDefinition blockDefinition, ApplicantData applicantData) {
+    return new AutoValue_Block(id, blockDefinition, applicantData);
+  }
 
   /**
    * The block's ID. Note this is different from the {@code BlockDefinition}'s ID because
    * BlockDefinitions that repeat expand to multiple Blocks.
    */
-  private final long id;
+  public abstract long id();
 
-  private final BlockDefinition blockDefinition;
-  private final ApplicantData applicantData;
-  // TODO: Make Block an AutoValue instead of implementing our own memoization.
-  private Optional<ImmutableList<ApplicantQuestion>> questionsMemo = Optional.empty();
-  private Optional<Boolean> errorsMemo = Optional.empty();
+  abstract BlockDefinition blockDefinition();
 
-  Block(long id, BlockDefinition blockDefinition, ApplicantData applicantData) {
-    this.id = id;
-    this.blockDefinition = checkNotNull(blockDefinition);
-    this.applicantData = checkNotNull(applicantData);
-  }
-
-  public long getId() {
-    return id;
-  }
+  abstract ApplicantData applicantData();
 
+  @Memoized
   public String getName() {
-    return blockDefinition.name();
+    return blockDefinition().name();
   }
 
+  @Memoized
   public String getDescription() {
-    return blockDefinition.description();
+    return blockDefinition().description();
   }
 
+  @Memoized
   public ImmutableList<ApplicantQuestion> getQuestions() {
-    if (questionsMemo.isEmpty()) {
-      this.questionsMemo =
-          Optional.of(
-              blockDefinition.programQuestionDefinitions().stream()
-                  .map(ProgramQuestionDefinition::getQuestionDefinition)
-                  .map(
-                      questionDefinition ->
-                          new ApplicantQuestion(questionDefinition, applicantData))
-                  .collect(toImmutableList()));
-    }
-    return questionsMemo.get();
+    return blockDefinition().programQuestionDefinitions().stream()
+        .map(ProgramQuestionDefinition::getQuestionDefinition)
+        .map(questionDefinition -> new ApplicantQuestion(questionDefinition, applicantData()))
+        .collect(toImmutableList());
   }
 
+  @Memoized
   public boolean hasErrors() {
-    if (errorsMemo.isEmpty()) {
-      this.errorsMemo =
-          getQuestions().stream().map(ApplicantQuestion::hasErrors).reduce(Boolean::logicalAnd);
-    }
-    return errorsMemo.get();
+    Optional<Boolean> hasErrors =
+        getQuestions().stream().map(ApplicantQuestion::hasErrors).reduce(Boolean::logicalAnd);
+    return hasErrors.isPresent() ? hasErrors.get() : false;
+  }
+
+  /**
+   * Checks whether the block is complete - that is, {@link ApplicantData} has values at all the
+   * paths for all required questions in this block. Note: this cannot be memoized, since we need to
+   * reflect internal changes to ApplicantData.
+   */
+  public boolean isComplete() {
```

Comment:
(question) if this cannot be memoized, then is it also true that `hasErrors` cannot be memoized also?

related: I think `isComplete` also needs to check `hasErrors`. If all the values are present but there are errors then this block is not complete.

If we ever need an `isFilledOut` method for other reasons, you could make an `isFilledOut` method, which looks like this one, and an `isComplete` that is `isFilledOut() && !hasErrors()`.

Reply:
Good point, I removed memoization from hasErrors and updated `isComplete` to `isCompleteWithoutErrors`